pkg/controllers/management/nodetemplate/nodetemplate.go

```diff
@@ -0,0 +1,236 @@
+package nodetemplate
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/rancher/rancher/pkg/controllers/management/rbac"
+	"github.com/rancher/rancher/pkg/namespace"
+	v1 "github.com/rancher/types/apis/core/v1"
+	v3 "github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/rancher/types/config"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+)
+
+const (
+	ownerBindingsAnno = "ownerBindingsCreated"
+)
+
+type nodeTemplateController struct {
+	ntClient        v3.NodeTemplateInterface
+	ntLister        v3.NodeTemplateLister
+	npLister        v3.NodePoolLister
+	npClient        v3.NodePoolInterface
+	nsLister        v1.NamespaceLister
+	nsClient        v1.NamespaceInterface
+	nodesLister     v3.NodeLister
+	nodeClient      v3.NodeInterface
+	ntDynamicClient dynamic.NamespaceableResourceInterface
+	mgmtCtx         *config.ManagementContext
+}
+
+func Register(ctx context.Context, mgmt *config.ManagementContext) {
+	// Parts of the node template in k8s are dynamic, such as an azureConfig. If the dynamic client is not use then the
+	// dynamic fields, like azureConfig, are stripped when marshalling the data into the NodeTemplate struct. If changes
+	// are made, and that struct is committed to k8s, then the dynamic field is effectively removed.
+	restConfig := mgmt.RESTConfig
+	dynamicClient, err := dynamic.NewForConfig(&restConfig)
+	if err != nil {
+		panic(fmt.Sprintf("[NodeTemplate Register] error creating dynamic client: %v", err))
+	}
+
+	s := schema.GroupVersionResource{
+		Group:    "management.cattle.io",
+		Version:  "v3",
+		Resource: "nodetemplates",
+	}
+	ntDynamicClient := dynamicClient.Resource(s)
+
+	nt := nodeTemplateController{
+		ntClient:        mgmt.Management.NodeTemplates(""),
+		ntLister:        mgmt.Management.NodeTemplates("").Controller().Lister(),
+		npLister:        mgmt.Management.NodePools("").Controller().Lister(),
+		npClient:        mgmt.Management.NodePools(""),
+		nsLister:        mgmt.Core.Namespaces("").Controller().Lister(),
+		nsClient:        mgmt.Core.Namespaces(""),
+		nodesLister:     mgmt.Management.Nodes("").Controller().Lister(),
+		nodeClient:      mgmt.Management.Nodes(""),
+		ntDynamicClient: ntDynamicClient,
+		mgmtCtx:         mgmt,
+	}
+
+	mgmt.Management.NodeTemplates("").Controller().AddHandler(ctx, "nt-grb-handler", nt.sync)
+}
+
+func (nt *nodeTemplateController) sync(key string, nodeTemplate *v3.NodeTemplate) (runtime.Object, error) {
+	if nodeTemplate == nil || nodeTemplate.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	annotations := nodeTemplate.Annotations
```

Comment:
Move the annotation check up to here `if nodeTemplate.Annotations !=nil && annotations[ownerBindingsAnno] == "true"`